crates/sui-transaction-builder/src/lib.rs

```diff
@@ -162,15 +164,22 @@ impl TransactionBuilder {
             .into_iter()
             .map(|id| self.get_object_ref(id))
             .collect();
-        let coins: Vec<ObjectRef> = join_all(handles)
-            .await
-            .into_iter()
-            .map(|c| c.unwrap())
-            .collect();
+        let coins: Vec<anyhow::Result<ObjectRef>> = join_all(handles).await.into_iter().collect();
+
+        if coins.iter().any(|c| c.is_err()) {
+            return Err(anyhow!("Failed reading PaySui input coin objects!"));
+        }
+        // unwrap() is safe here because in case of err, the function would have returned err already.
+        let coin_refs: Vec<ObjectRef> = coins.into_iter().map(|c| c.unwrap()).collect();
         // [0] is safe because input_coins is non-empty and coins are of same length as input_coins.
-        let gas_object = *coins.get(0).unwrap();
+        let gas_object_ref = coin_refs[0];
```

Comment:
not sure if I followed, `gas_object_ref` here only guaranteed to be of ObjectRef type, coin type-check will happen in https://github.com/MystenLabs/sui/blob/88197f2325f6ee6a97a267e7b897ad73562f71ca/crates/sui-core/src/execution_engine.rs#L324

Reply:
right, I'm wondering if we should do the early check here

Reply:
We cannot move the checks here as TS SDK and manually constructed tx will not go thru the tx builder route here.
Duplicating the checks would catch the errs here instead of around input_checker, but it also requires reading and parsing the coin objs rather than the obj refs, so I did not see a lot of values duplicating that, wdyt?